### ISSUE TITLE
strange api failure

### ISSUE THREAD

**Issue by whyrusleeping** [ipfs/go-ipfs#1620]
something strange appears to have regressed:
```
whyrusleeping@idril ~> ipfs add bigfile2
added Qmd1usDGx7zDU5zq3M4WQft6ENVQJtzc8XNYy4qzPEgKWE bigfile2
whyrusleeping@idril ~> cat bigfile2 | shasum
5ab871067c871695e1d7f99548f451aed6726e9f  -
whyrusleeping@idril ~> ipfs cat Qmd1usDGx7zDU5zq3M4WQft6ENVQJtzc8XNYy4qzPEgKWE | shasum
Error: read tcp 127.0.0.1:58756->127.0.0.1:5001: read: connection reset by peer
e788ad60376eab9962dbaeac0240dcd87c83c4b6  -
```

bigfile2 is exactly 100MB, random data from /dev/urandom

**Comment by whyrusleeping** [comment]
failing commit: 
```
a7e50f1fbc5a58963a1b1e462575249de61540f1 is the first bad commit
commit a7e50f1fbc5a58963a1b1e462575249de61540f1
Author: Jeromy <jeromyj@gmail.com>
Date:   Fri Jul 24 17:41:59 2015 -0700

    implement http trailers for errors after headers are sent
    
    refactor http handler and copyChunks to get this all to work correctly
    License: MIT
    Signed-off-by: Jeromy <jeromyj@gmail.com>

:040000 040000 4bbe16ab19466d3ebca59863f6ed19e58cd5fa5a aa4a0acff1d9eb6c1db6079873aaaab67b62579b M	commands

```

**Comment by whyrusleeping** [comment]
The issue is that the final `Read` call on the client gets valid data *and* returns an error from the other side closing the connection.

**Comment by whyrusleeping** [comment]
i suspect this is a go1.5 issue.

**Comment by whyrusleeping** [comment]
fixed in #1621
